cmd/rukpakctl/cmd/run.go

```diff
@@ -0,0 +1,182 @@
+  bundle "memcached-api-8578dfddf9" is already up-to-date
+`,
+		Args: cobra.ExactArgs(2),
+		Run: func(cmd *cobra.Command, args []string) {
+			bundleDeploymentName, bundleDir := args[0], args[1]
+			ctx := signals.SetupSignalHandler()
+
+			cfg := ctrl.GetConfigOrDie()
+
+			sch := scheme.Scheme
+			if err := rukpakv1alpha1.AddToScheme(sch); err != nil {
+				log.Fatal(err)
+			}
+			cl, err := client.New(cfg, client.Options{Scheme: sch})
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			bundleFS := os.DirFS(bundleDir)
+			digest := fnv.New64a()
+			if err := util.FSToTarGZ(digest, bundleFS); err != nil {
+				log.Fatal(err)
+			}
+
+			bundleLabels := map[string]string{
+				"app":          bundleDeploymentName,
+				"bundleDigest": fmt.Sprintf("%x", digest.Sum(nil)),
+			}
+
+			bd := buildBundleDeployment(bundleDeploymentName, bundleLabels, bundleDeploymentProvisionerClassName, bundleProvisionerClassName)
+			if err := cl.Patch(ctx, bd, client.Apply, client.ForceOwnership, client.FieldOwner("rukpakctl")); err != nil {
+				log.Fatalf("apply bundle deployment: %v", err)
+			}
+			fmt.Printf("bundledeployment.core.rukpak.io %q applied\n", bundleDeploymentName)
+
+			rukpakCAs, err := rukpakctl.GetClusterCA(ctx, cl, systemNamespace, caSecretName)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			bundleName, err := getBundleName(ctx, cfg, bundleLabels)
+			if err != nil {
+				log.Fatalf("failed to get bundle name: %v", err)
+			}
+			bu := rukpakctl.BundleUploader{
+				UploadServiceName:      uploadServiceName,
+				UploadServiceNamespace: systemNamespace,
+				Cfg:                    cfg,
+				RootCAs:                rukpakCAs,
+			}
+			modified, err := bu.Upload(ctx, bundleName, bundleFS)
+			if err != nil {
+				log.Fatalf("failed to upload bundle: %v", err)
+			}
+			if !modified {
+				fmt.Printf("bundle %q is already up-to-date\n", bundleName)
+				return
+			}
+
+			fmt.Printf("successfully uploaded bundle content for %q\n", bundleName)
+		},
+	}
+	cmd.Flags().StringVar(&systemNamespace, "system-namespace", "rukpak-system", "the namespace in which the rukpak controllers are deployed.")
+	cmd.Flags().StringVar(&uploadServiceName, "upload-service-name", "upload-manager", "the name of the service of the upload manager.")
+	cmd.Flags().StringVar(&caSecretName, "ca-secret-name", "rukpak-ca", "the name of the secret in the system namespace containing the root CAs used to authenticate the upload service.")
+	cmd.Flags().StringVar(&bundleDeploymentProvisionerClassName, "bundle-deployment-provisioner-class", "core.rukpak.io/plain", "Provisioner class name to set on bundle deployment.")
+	cmd.Flags().StringVar(&bundleProvisionerClassName, "bundle-provisioner-class", "core.rukpak.io/plain", "Provisioner class name to set on bundle.")
+	return cmd
+}
+
+func buildBundleDeployment(bdName string, bundleLabels map[string]string, biPCN, bPNC string) *unstructured.Unstructured {
+	return &unstructured.Unstructured{Object: map[string]interface{}{
```

Comment:
What's the benefit of not using a typed BD object here?

Reply:
If you use a typed object, any default values from that struct get serialized into the JSON patch, which could cause unrelated fields to be patched back to the default value even though that isn't the intention. Using an unstructured ensures that the patch contains _only_ what is specified.

Reply:
That seems like a reasonable explanation to me. Does it make sense to capture that as a comment there?